src/map.rs

```diff
@@ -260,7 +265,7 @@ impl<Sz> ShortHashProxy<Sz>
 /// assert_eq!(letters.get(&'y'), None);
 /// ```
 #[derive(Clone)]
-pub struct IndexMap<K, V, S = RandomState> {
+pub struct IndexMap<K, V, S = DefaultHashBuilder> {
```

Comment:
Oh, this is a breaking change though, and `RandomState` is only defined in `std`.

Reply:
Is there any scenario in which this causes problems? Type inference should solve it in nearly all cases; I'd be surprised if there's anything that would actually depend on type inference guessing the type of the hasher.

Reply:
It's debatable how likely this is, but if someone was explicit about `S` in one place and assumed the default elsewhere, it would break. Something like:

```rust
type MyMap = IndexMap<MyKey, MyValue, RandomState>; // explicit `S`

fn create_my_map(args...) -> MyMap {
    let mut map = IndexMap::new(); // this uses the default `S`!
    // fill the map
    // ...
    map
}
```

That's a direct breaking change, where something would stop compiling.

There's also the softer semantic change to consider in DOS-resistance, from `std`'s SipHash to `hashbrown`'s choice of aHash. That does still claim to be DOS-resistant, but I don't know enough to evaluate this. If it's really all beneficial, faster and still secure, then aHash should be advocated for `std` itself, and I'd feel better about following suit.

Reply:
aHash only supports x86 and ARM. hashbrown falls back to FxHash on other architectures.

Reply:
Ah, then I'm even more wary of this change.

Reply:
The `siphash` crate provides a `no_std` implementation of SipHash. Should I add a dependency on it?

Reply:
That crate looks abandoned -- I'd be more inclined to roll our own internally. It also doesn't provide a `BuildHasher`, which AIUI needs randomization to be DOS-resistant.

Changing from `RandomState` is still a breaking change at the type level though.

Reply:
A possible way out -- detect the availability of `std` in the build script. If present, we can use `S = RandomState`, otherwise I'd leave it with no default at all. The default-`S` methods `new` and `with_capacity` would also be gated entirely.

So we'd only truly be `no_std` for crates that don't have `std` at all, but I think that's enough. We don't support those targets at all right now, so it's not a breaking change.